### ISSUE TITLE
Blocked on alloc under macOS

### ISSUE THREAD

**Issue by erikgrinaker** [cockroachdb/pebble#1111]
I'm running a local roachprod cluster on macOS with 8 nodes (cockroachdb/cockroach@ce3b29b71f), and creating a table with some data hangs forever:

```
$ roachprod create local -n 8
$ roachprod start local --racks 3
```

```sql
CREATE DATABASE test;
ALTER DATABASE test CONFIGURE ZONE USING num_replicas = 5, range_min_bytes = 1e6, range_max_bytes=10e6;
USE test;
CREATE TABLE data AS SELECT id, REPEAT('x', 1024) FROM generate_series(1, 1e6) AS id;
```

The logs kept repeating these messages over and over:

```
E210407 14:01:42.440401 2322884 kv/kvserver/queue.go:1093 ⋮ [n6,raftsnapshot,s6,r78/2:‹/{Table/53/1/-…-Max}›] 5294  (n8,s8):3: remote couldn't accept VIA_SNAPSHOT_QUEUE snapshot ‹c40b7a30› at applied index 13 with error: ‹[n8,s8],r78: cannot apply snapshot: snapshot intersects existing range; initiated GC: [n8,s8,r69/3:/{Table/53/1/-…-Max}] (incoming /{Table/53/1/-9222246136947506714-Max})›
```

Range 69 is stuck applying Raft entries before a range split, which is why the GC never goes through:

<img width="847" alt="Screenshot 2021-04-07 at 16 06 09" src="https://user-images.githubusercontent.com/644420/113879837-380d3400-97bb-11eb-8332-d875b777230a.png">

Looking at the goroutine stacks, we found this Pebble call which has been blocked on alloc for 67 minutes:

```
goroutine 250 [syscall, 67 minutes]:
github.com/cockroachdb/pebble/internal/manual._Cfunc_calloc(0x79, 0x1, 0x0)
	_cgo_gotypes.go:42 +0x49
github.com/cockroachdb/pebble/internal/manual.New(0x79, 0x73, 0x1056f1, 0x0)
	/Users/erik/Projects/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/internal/manual/manual.go:40 +0x3d
github.com/cockroachdb/pebble/internal/cache.newValue(0x59, 0x2)
	/Users/erik/Projects/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/internal/cache/value_normal.go:38 +0x38
github.com/cockroachdb/pebble/internal/cache.(*Cache).Alloc(...)
	/Users/erik/Projects/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/internal/cache/clockpro.go:696
github.com/cockroachdb/pebble/sstable.(*Reader).readBlock(0xc001e28a00, 0x1056f1, 0x54, 0x0, 0x0, 0x3fd9384950000000, 0x6, 0x7)
	/Users/erik/Projects/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/sstable/reader.go:1910 +0x165
github.com/cockroachdb/pebble/sstable.(*Reader).readMetaindex(0xc001e28a00, 0x1056f1, 0x54, 0x0, 0x0)
	/Users/erik/Projects/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/sstable/reader.go:2012 +0x7d
github.com/cockroachdb/pebble/sstable.NewReader(0x13c2d040, 0xc0004c6698, 0xc000c62460, 0xc5590a0, 0xc0014b5560, 0x8718864, 0x18, 0xc001cc2440, 0x1, 0x1, ...)
	/Users/erik/Projects/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/sstable/reader.go:2309 +0x348
github.com/cockroachdb/pebble.ingestLoad1(0xc0012a01e0, 0xc0016359a0, 0x49, 0x2, 0x73, 0x0, 0x0, 0x0)
	/Users/erik/Projects/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/ingest.go:60 +0x2cd
github.com/cockroachdb/pebble.ingestLoad(0xc0012a01e0, 0xc0025f6ae0, 0x1, 0x1, 0x2, 0xc00381d718, 0x1, 0x1, 0xc001cc25c0, 0x10000c001cc25f0, ...)
	/Users/erik/Projects/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/ingest.go:161 +0x1a5
github.com/cockroachdb/pebble.(*DB).Ingest(0xc0007e0400, 0xc0025f6ae0, 0x1, 0x1, 0x1, 0xc0025f6ae0)
	/Users/erik/Projects/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/ingest.go:528 +0x19e
github.com/cockroachdb/cockroach/pkg/storage.(*Pebble).IngestExternalFiles(0xc001281380, 0x9bbdf50, 0xc0026774d0, 0xc0025f6ae0, 0x1, 0x1, 0x0, 0x0)
	/Users/erik/Projects/go/src/github.com/cockroachdb/cockroach/pkg/storage/pebble.go:1110 +0x4c
github.com/cockroachdb/cockroach/pkg/kv/kvserver.addSSTablePreApply(0x9bbdf50, 0xc0026774d0, 0xc000ec0000, 0x9d0ad68, 0xc001281380, 0x9c08d48, 0xc001222580, 0x7, 0x27, 0xc004cf4000, ...)
	/Users/erik/Projects/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_proposal.go:577 +0x762
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaAppBatch).runPreApplyTriggersAfterStagingWriteBatch(0xc000b43be0, 0x9bbdf50, 0xc0026774d0, 0xc00106e008, 0x0, 0x0)
	/Users/erik/Projects/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_application_state_machine.go:616 +0xcfc
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaAppBatch).Stage(0xc000b43be0, 0x9bbe6f8, 0xc00106e008, 0xc001cc3338, 0x5ae3c4b, 0xc000b43df8, 0xc000b43e28)
	/Users/erik/Projects/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_application_state_machine.go:507 +0x3be
github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply.mapCmdIter(0x9c005c0, 0xc000b43df8, 0xc001cc3470, 0x0, 0x0, 0x0, 0x0)
	/Users/erik/Projects/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply/cmd.go:175 +0x142
github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply.(*Task).applyOneBatch(0xc001cc3958, 0x9bbdf50, 0xc0026774d0, 0x9c005c0, 0xc000b43dc8, 0x0, 0x0)
	/Users/erik/Projects/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply/task.go:280 +0x185
github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply.(*Task).ApplyCommittedEntries(0xc001cc3958, 0x9bbdf50, 0xc0026774d0, 0x1, 0x8760951)
	/Users/erik/Projects/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply/task.go:247 +0xc5
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReadyRaftMuLocked(0xc000b43b00, 0x9bbdf50, 0xc0026774d0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/erik/Projects/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_raft.go:796 +0xfad
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReady(0xc000b43b00, 0x9bbdf50, 0xc0026774d0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/erik/Projects/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_raft.go:459 +0x113
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processReady(0xc000e72a00, 0x9bbdf50, 0xc000f1ff20, 0x45)
	/Users/erik/Projects/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_raft.go:523 +0x134
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).worker(0xc001854b40, 0x9bbdf50, 0xc000f1ff20)
	/Users/erik/Projects/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/scheduler.go:284 +0x2c2
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1(0xc0014c0580, 0x9bbdf50, 0xc000f1ff20, 0xc000e83340, 0xc001ad8e70)
	/Users/erik/Projects/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:351 +0xb9
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask
	/Users/erik/Projects/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:346 +0xfc
```

**Comment by petermattis** [comment]
That's quite bizarre. I can't imagine why a call to `malloc` would block for so long. Was the process memory usage right at the limits of the machine's memory capacity?

**Comment by erikgrinaker** [comment]
No, I have 12 GB free memory. Will see if I can find any process limits or such.

**Comment by petermattis** [comment]
Ack. That makes no sense. Maybe there is something Mac OS X specific.

**Comment by erikgrinaker** [comment]
Haven't found any clues, but it does appear to be fairly reproducible since I now have 4 out of 8 nodes with Goroutines blocked on this. It's probably some sort of macOS resource limit that kicks in. Will try to run a few more clusters and see if it crops up again.

**Comment by petermattis** [comment]
Running 8 nodes on your laptop is a bit stressful. I'd be interested to know if you can reproduce this on a linux box.

**Comment by petermattis** [comment]
Another thought: if you can attach a debugger to one of these blocked nodes, perhaps you can see what is going on in the C call. I haven't tried that myself on Mac OS X lately, so I'm not sure what state the tooling is in.

**Comment by erikgrinaker** [comment]
> I'd be interested to know if you can reproduce this on a linux box.

Nope, ran 24-node clusters with 10 replicas on a single GCE worker just fine.

> if you can attach a debugger to one of these blocked nodes, perhaps you can see what is going on in the C call

```
(lldb) process attach --pid 83572
Process 83572 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
    frame #0: 0x00007fff205428e2 libsystem_kernel.dylib`__psynch_cvwait + 10
libsystem_kernel.dylib`__psynch_cvwait:
->  0x7fff205428e2 <+10>: jae    0x7fff205428ec            ; <+20>
    0x7fff205428e4 <+12>: movq   %rax, %rdi
    0x7fff205428e7 <+15>: jmp    0x7fff205406a1            ; cerror_nocancel
    0x7fff205428ec <+20>: retq   
  thread #14, stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
    frame #0: 0x00007fff2a73d8b0 libunwind.dylib`libunwind::CFI_Parser<libunwind::LocalAddressSpace>::decodeFDE(libunwind::LocalAddressSpace&, unsigned long, libunwind::CFI_Parser<libunwind::LocalAddressSpace>::FDE_Info*, libunwind::CFI_Parser<libunwind::LocalAddressSpace>::CIE_Info*) + 30
libunwind.dylib`libunwind::CFI_Parser<libunwind::LocalAddressSpace>::decodeFDE:
->  0x7fff2a73d8b0 <+30>: movl   (%rsi), %r13d
    0x7fff2a73d8b3 <+33>: leaq   0x4(%rsi), %rbx
    0x7fff2a73d8b7 <+37>: movl   $0xffffffff, %eax         ; imm = 0xFFFFFFFF 
    0x7fff2a73d8bc <+42>: movq   %rbx, -0x30(%rbp)
Target 0: (cockroach) stopped.
```

This might be an actual illegal memory use exception, that ends up blocking the goroutine rather than panicing. Unless this is somehow an artifact of Cgo or debugger/Go interactions, but I only saw this on processes that had goroutines blocked on `calloc`.

Running as root didn't make any difference, neither did bumping all `ulimit` limits (though I didn't really expect this to matter anyway). I'm able to reproduce this in 100% of runs, for about half of the nodes.

**Comment by petermattis** [comment]
> This might be an actual illegal memory access exception, that ends up blocking the goroutine rather than panicing. Unless this is somehow an artifact of Cgo or debugger/Go interactions, but I only saw this on processes that had goroutines blocked on calloc.

I'd be very surprised if an illegal memory access didn't kill the process. I suspect that the `EXC_BAD_ACCESS` message is an artifact of using lldb on a Go binary. Have you tried using [`delve`](https://github.com/go-delve/delve)? I've heard it is pretty decent at this stuff nowadays.

**Comment by knz** [comment]
Are you running falcond? This may be a protection of falcond against programs that use too much resources.

**Comment by erikgrinaker** [comment]
> Are you running falcond? This may be a protection of falcond against programs that use too much resources.

I am indeed, could very well be something like that. Will check the falcond logs next time I try to repro this, thanks!

**Comment by erikgrinaker** [comment]
Been poking at this a bit more, and not really getting anywhere -- it's the weirdest thing. I'm starting to think I must be missing something embarrassingly obvious.

It's not falcond, since it happens on my personal iMac as well. I've freed up tons of memory, even on my 64 GB iMac, still happens (and the processes aren't using much memory anyway, less that 1 GB each). The blocked allocations are small, like ~1500b.

Tried out DTrace, and traced every single `calloc` call and return from all processes. They all matched up, i.e. I didn't find any `libsystem_malloc.dylib:calloc` calls that didn't return. I suppose that's reassuring. The processes don't receive any CPU faults either, so the `EXC_BAD_ACCESS` was bogus as expected.

`delve` doesn't get past `runtime.asmcgocall`, which has switched to the system stack:

```
> runtime.asmcgocall() /usr/local/Cellar/go/1.16.2/libexec/src/runtime/asm_amd64.s:653 (PC: 0x4082482)
Warning: debugging optimized function
   648:		JEQ	nosave
   649:	
   650:		// Switch to system stack.
   651:		MOVQ	m_g0(R8), SI
   652:		CALL	gosave<>(SB)
=> 653:		MOVQ	SI, g(CX)
   654:		MOVQ	(g_sched+gobuf_sp)(SI), SP
   655:	
   656:		// Now on a scheduling stack (a pthread-created stack).
   657:		// Make sure we have enough room for 4 stack-backed fast-call
   658:		// registers as per windows amd64 calling convention.
```

However, every once in a blue moon the goroutine stack trace will actually return stuff from beyond the Cgo boundary:

```
goroutine 27264 [syscall, 1 minutes]:
pthread_getspecific
	/usr/lib/system/libsystem_pthread.dylib:0 pc=0x7fff205ed4e6
je_arena_malloc_large
	/Users/erik/Projects/go/src/github.com/cockroachdb/cockroach/cockroach:0 pc=0x7bd3609
je_calloc
	/Users/erik/Projects/go/src/github.com/cockroachdb/cockroach/cockroach:0 pc=0x7bbea9a
_malloc_zone_calloc
	/usr/lib/system/libsystem_malloc.dylib:0 pc=0x7fff20431eb7
_cgo_82be0c43bdb2_Cfunc_calloc
	/Users/erik/Projects/go/src/github.com/cockroachdb/cockroach/cockroach:0 pc=0x7ac9991
github.com/cockroachdb/pebble/internal/manual._Cfunc_calloc(0x7c1, 0x1, 0x0)
	_cgo_gotypes.go:45 +0x4e
github.com/cockroachdb/pebble/internal/manual.New(0x7c1, 0x867ca00, 0x191900, 0x112d63b0)
/Users/erik/Projects/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/pebble/internal/manual/manual.go:40 +0x3d
```

`pthread` could conceivably explain why things would block. This reminded me that CRDB uses jemalloc, so I did a `stdmalloc` build, but it had the same problem. I can't help but wonder whether this could be a Go scheduler issue or something.

I'm not really sure where to go from here. Are there any low-level hackers around that we could nerdsnipe?

**Comment by erikgrinaker** [comment]
Delve interestingly has a different state `GC assist marking` for the hung goroutines (weird timestamp though):

```
Goroutine 11112 - User: _cgo_gotypes.go:42 github.com/cockroachdb/pebble/internal/manual._Cfunc_calloc (0x595af69) [GC assist marking 449009h4m5.61671247s]
```

Tried setting `GOGC=off`, but still saw hung goroutines. Did a few runs where I disabled the Pebble CGo memory allocation by setting `cgoEnabled=false`, which didn't show any problems apart from being slow.

I'll try to come up with a minimal Pebble-only repro case for this, and see if the Go runtime diagnostics show anything useful.

**Comment by petermattis** [comment]
I tried reproducing this today, but have so far been unable. After running for ~10 min, I only see 3 "remote couldn't accept VIA_SNAPSHOT_QUEUE snapshot" messages in the logs. I checked the goroutine dumps on each node and was unable to locate one that has a goroutine blocked in `_Cfunc_calloc`. I'm going to let this continue to run in case it takes longer than 10 min to reproduce.

**Comment by petermattis** [comment]
Huh, the `CREATE TABLE ... AS SELECT` finished:

```
root@localhost:26257/test> CREATE TABLE data AS SELECT id, REPEAT('x', 1024) FROM generate_series(1, 1e6) AS id;

CREATE TABLE AS

Time: 776.941s total (execution 0.935s / network 776.006s)
```

**Comment by erikgrinaker** [comment]
I only check for hung goroutines, the first one usually happens about 2-3 minutes in, e.g.:

```
for URL in $(roachprod adminurl local); do echo $URL; curl -sSf "${URL}debug/pprof/goroutine?debug=2" | grep -B 20 '.Alloc('; done
```

The create will sometimes go through, it depends on which goroutines gets blocked.

Running Big Sur and Go 1.16, for reference.

**Comment by petermattis** [comment]
I'm running on Catalina with go1.15.8. I'm trying to reproduce again with Catalina on go1.16.2. Given the oddity in what you're seeing, I could very well believe it is specific to go1.16, or to Big Sur.

**Comment by petermattis** [comment]
Just ran two attempts at reproduction with a CRDB (master) compiled with go1.16.2. Both attempts ran the CTAS operation to completion and the per-node goroutine output revealed no hung goroutines.

**Comment by erikgrinaker** [comment]
Interesting, thanks for testing! Might be a Big Sur thing, or something even more env-specific. Will try to get a repro on a different system.

**Comment by erikgrinaker** [comment]
@stevendanna was able to reproduce this on Big Sur. Interestingly, he accidentally run with an older binary first that didn't have this problem, so I suppose that gives us something to bisect by:

```
CockroachDB CCL v20.2.4 (x86_64-apple-darwin14, built 2021/01/21 00:12:56, go1.13.14)
```

Will try a few different versions of CRDB and Go.

**Comment by erikgrinaker** [comment]
It's Go 1.14.

| CRDB | Go | Behavior |
|--|--|--|
| v20.2.7 download | 1.13.14 | ✅ |
| v21.1.0-beta.2 download | 1.15.11 | ❌ |
| master @ 84a84d56b3 | 1.16.3 | ❌ |
| master @ 84a84d56b3 | 1.15.11 | ❌ |
| master @ 84a84d56b3 | 1.13.15 | Won't compile |
| release-20.2 @ ae0d209448 | 1.16.3 | ❌ |
| release-20.2 @ ae0d209448 | 1.14.15 | ❌ |
| release-20.2 @ ae0d209448 | 1.13.15 | ✅ |

Release notes say:

>Goroutines are now asynchronously preemptible. As a result, loops without function calls no longer potentially deadlock the scheduler or significantly delay garbage collection. This is supported on all platforms except windows/arm, darwin/arm, js/wasm, and plan9/*.
>
>A consequence of the implementation of preemption is that on Unix systems, including Linux and macOS systems, programs built with Go 1.14 will receive more signals than programs built with earlier releases. This means that programs that use packages like syscall or golang.org/x/sys/unix will see more slow system calls fail with EINTR errors. Those programs will have to handle those errors in some way, most likely looping to try the system call again. For more information about this see man 7 signal for Linux systems or similar documentation for other systems.

This seems like it could cause what we're seeing, but I'm not sure if it's our responsibility or the runtime's in this case -- I don't believe `calloc` is a syscall, and the syscall the goroutine refers to is the Cgo switch to the system stack. In any case, I'll try to fiddle around with some signal handlers.

**Comment by petermattis** [comment]
> It's Go 1.14.

Yet I wasn't able to reproduce the problem using go1.15.8 on Catalina. Perhaps something changed in Big Sur. I just upgraded to Big Sur yesterday, so I might give this another go at reproducing today.

The async preemption can be disabled via `GODEBUG=asyncpreemptoff=1`:

>	asyncpreemptoff: asyncpreemptoff=1 disables signal-based
	asynchronous goroutine preemption. This makes some loops
	non-preemptible for long periods, which may delay GC and
	goroutine scheduling. This is useful for debugging GC issues
	because it also disables the conservative stack scanning used
	for asynchronously preempted goroutines.

**Comment by erikgrinaker** [comment]
> Yet I wasn't able to reproduce the problem using go1.15.8 on Catalina. Perhaps something changed in Big Sur.

Must be a combination.

> The async preemption can be disabled via `GODEBUG=asyncpreemptoff=1`:

Nice find, this resolves the problem. CRDB also appears to run a lot less hot (i.e. my fans no longer turn on).

**Comment by petermattis** [comment]
> Nice find, this resolves the problem. CRDB also appears to run a lot less hot (i.e. my fans no longer turn on).

🎉 

I also just noticed that my CRDB was consuming 100% of CPU when running on Big Sur. Disabling async preempt "fixed" the issue. Have you searched the issues at github.com/golang/go to see if there are any other reports of problems with async preemption and Big Sur?

**Comment by erikgrinaker** [comment]
> Have you searched the issues at github.com/golang/go to see if there are any other reports of problems with async preemption and Big Sur?

Briefly, but I'll have a closer look. I think this is probably a combination of Cgo/Pebble and async preemption though, since I haven't seen anything else get stuck in this way.

**Comment by erikgrinaker** [comment]
I haven't been able to come up with a reduced test case for this, having tried various variations of concurrent stress tests of the `manual` package, `cache` package, and Pebble as a whole. Haven't found any solid clues as to what causes these hangs either. I've also tried out custom Go builds with changes to the `darwin`-specific code in the scheduler, but no luck.

I'm going to leave this for a bit to help out with some release blockers, since this doesn't seem like a critical issue, but I've opened https://github.com/golang/go/issues/45558 with what we know so far. I'll follow up on this and keep working on a reduced test case in between.
